### PR TITLE
CDash integration plus coverage and sanitizers support

### PR DESCRIPTION
**Pull request by tcojean**
Provide support in Ginkgo for Coverage and Sanitizers build and CDash.

+ Add a CMake script which simplifies adding new build types to the project
+ Use this script to provide three new build types:
    + COVERAGE: relies on gcov and the `--coverage` flag
    + ASAN: relies on the flag `-fsanitize=address`
    + TSAN: relies on the flag `-fsanitize=thread`
+ Add four new steps to the CI system, one for each of the new build types plus valgrind.
+ Update CI benchmarking which was outdated.
+ Add configuration files to simplify using CTest and support CDash.
+ Document these additions in the Readme.

### Requirements
+ [x] Merge PR #220 
+ [x] Rebase this PR

### Things to discuss
There are multiple things to discuss, but I want to raise some particular points.

This PR should make each merge into develop/master take one to two hours. Thankfully, the most important steps (sync, build, test, generate doc) should be done fairly fast. After that comes the slow stuff which helps us gathering data. These would be:
+ Coverage,
+ Valgrind run,
+ ThreadSanitizer run,
+ AdressSanitizer run.

All these data would nicely be available in our CDash dashboard to actually track progress (or regression) in all of those fields. Would you think this is too much?

### CODE REVIEW

**Comment by tcojean** [comment]
In addition, I would like to add CUDA benchmarking for every PR merged into master or develop. This means that every merge would imply fairly long benchmarking on top of all of this (I don't remember the exact numbers). I think it's important to always run benchmarks when merging to master or develop because that is the only way to properly track performance changes.

For that to work, I think we **really** need to come up with a small-ish (200? 500?) set of matrices we think are reliable enough but fast enough to run all the time. If we do, we should be able to shrink the benchmarking time by quite a bit. It would be interesting to add OpenMP benchmarks but this would require more isolation, our CI system is too busy for that.

**Comment by tcojean** [comment]
You can look at this pipeline (currently running) for an estimate of how much time a full run can take:
https://gitlab.com/ginkgo-project/ginkgo-public-ci/pipelines/47996433

An example of output for this run will be available on the dashboard, which I made public now that the setup seems stable enough:
https://my.cdash.org/index.php?project=Ginkgo+Project&date=2019-02-18

A previous run, although incomplete as there was trouble with Valgrind and ASAN, is available on this date:
https://my.cdash.org/index.php?project=Ginkgo+Project&date=2019-02-16

**Comment by tcojean** [comment]
This can now be reviewed.

**Comment by pratikvn** [comment]
This is definitely a very good idea. But I see that it takes about 2.5 hours approx. I see that the [split](https://gitlab.com/ginkgo-project/ginkgo-public-ci/pipelines/47996433/builds) is 10mins, 1.25 hours, 45 mins, 15 mins for coverage, valgrind, Tsan and Asan respectively.

As expected, valgrind is much slower. As [this](https://github.com/google/sanitizers/wiki/AddressSanitizerComparisonOfMemoryTools) comparison shows, Asan+Tsan (+ maybe Msan) should be faster/better than valgrind ? valgrind+Tsan+Asan seems to be overkill. 

One comment: The overhead of the thread sanitizer is higher than that of the address sanitizer right ? So, maybe we can run Asan first and make Tsan optional ?

File: README.md [hunk 1]
```diff
@@ -89,11 +89,15 @@ Ginkgo adds the following additional switches to control what is being built:
...
-*   `-DCMAKE_INSTALL_PREFIX=path` sets the installation path for `make install`.
-    The default value is usually something like `/usr/local`
 *   `-DGINKGO_VERBOSE_LEVEL=integer` sets the verbosity of Ginkgo.
     * `0` disables all output in the main libraries,
     * `1` enables a few important messages related to unexpected behavior (default).
+*   `-DCMAKE_INSTALL_PREFIX=path` sets the installation path for `make install`.
+    The default value is usually something like `/usr/local`
+*   `-DCMAKE_BUILD_TYPE=type` specifies which configuration will be used for
+    this build of Ginkgo. There are no default. Supported values are CMake's
+    standard build types such as `DEBUG` and `RELEASE` and the Ginkgo specific
```
Comment (pratikvn) [inline README.md]: What is the default value of this variable ?

Comment (tcojean) [reply README.md]: As stated, `there are no default`.

**Comment by tcojean** [comment]
> As expected, valgrind is much slower. As [this](https://github.com/google/sanitizers/wiki/AddressSanitizerComparisonOfMemoryTools) comparison shows, Asan+Tsan (+ maybe Msan) should be faster/better than valgrind ? valgrind+Tsan+Asan seems to be overkill.

I'm unsure about that in fact, there were problems with the container so the Asan run is maybe not complete. Technically, Asan is supposed to be faster than Tsan though (twice as much at least).

Furthermore, using Msan from what I understood is unrealistic. I believe it requires that all dependencies are built with Msan support, which should be impossible for CUDA etc.

**Comment by tcojean** [comment]
Another thing I need to look into for the Valgrind vs Asan thing, is that the Valgrind suppressions are easy to setup, I know it is possible for Asan but I need to check that in more details. I'm also unsure if Asan alone gives the same errors as Valgrind.

I will run Asan once more for this PR and we can check that.

**Comment by pratikvn** [comment]
> On another note, this is ran only with merged pull requests into master or develop, and while it does take a while, the amount of time is not very important as there is not much merged PRs per day, and even then these long jobs mostly take one to four CPUs for OpenMP (at least running the tests, which is the long part) so you could run multiple of them in parallel.

That actually sounds good. I guess I was thinking of maybe running Asan for every push to the repo, but that is unnecessary, now that I think about it.

File: README.md (hunk 1 above)
Comment (pratikvn) [reply README.md]: Maybe `RelWithDebInfo` as default is ideal ?

**Comment by pratikvn** [comment]
We get some [failed tests](https://my.cdash.org/viewTest.php?onlyfailed&buildid=1610960) which seems to be a bug with Asan. The [solution](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=58680) seems to be add the `-static-libasan` flag.

**Comment by gflegar** [comment]
> This is definitely a very good idea. But I see that it takes about 2.5 hours approx. I see that the [split](https://gitlab.com/ginkgo-project/ginkgo-public-ci/pipelines/47996433/builds) is 10mins, 1.25 hours, 45 mins, 15 mins for coverage, valgrind, Tsan and Asan respectively.

I would suggest the following:

- make these jobs manual for all feature branches
- run them nightly on the develop and master branches

This way, you remove the 2.5h merge delay, but you get the report every time something went wrong at least a day late, so you can go back and run the manual jobs for everything merged the previous day and figure out what caused the problem

**Comment by tcojean** [comment]
> This way, you remove the 2.5h merge delay, but you get the report every time something went wrong at least a day late, so you can go back and run the manual jobs for everything merged the previous day and figure out what caused the problem

To me, there is no real 2.5h merge delay there. In fact, once the `sync` step is ran the merge factually happened. Then after another 30 minutes you get the build results and the tests, so you are pretty certain everything went correctly. I don't really think it's a problem from then on that it takes an extra hour or two to get the green check mark.

Running this nightly is another option, but that sort of strays away from the "Continuous" aspect which I would rather not, for multiple reasons. One is that manually running separate tests on each MR to catch the problems is in fact a pain.

**Comment by tcojean** [comment]
> We get some [failed tests](https://my.cdash.org/viewTest.php?onlyfailed&buildid=1610960) which seems to be a bug with Asan. The [solution](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=58680) seems to be add the `-static-libasan` flag.

No the problem is in fact due to [this](https://stackoverflow.com/questions/49735926/address-sanitizer-with-gcc-fails-on-ubuntu-17-10-docker-container), but I'm supposed to have fixed this already....

File: README.md (hunk 1 above)
Comment (gflegar) [reply README.md]: There is a default, and default is "" (empty string), that is also an option, in which case none of the variables `CMAKE_SOMETHING_<CONFIG>` get merged into their `CMAKE_SOMETHING` variants.

Here are the default initializations for the GNU compiler (which are also inherited by the Clang compiler) for various configurations: https://github.com/Kitware/CMake/blob/master/Modules/Compiler/GNU.cmake#L50

Looking at these, `RelWithDebInfo` should not be interpreted as "a good default which provides both optimization and debug capabilities", but rather as "an option to debug problems that arise from enabling optimizations", which is probably not something very useful for the average user. I would go for `Release` as the default, since that is what will be the first thing users want to try - how fast does this thing work, and instruct them to build and link with the debug version if they think they encountered a bug.

Comment (tcojean) [reply README.md]: Thanks for looking, I did know of empty (`""`) but was never sure what it meant. So basically by default there are no particular flags specified, is that correct?

I agree with making Release default. I will do this as part of this PR.

**Comment by gflegar** [comment]
> To me, there is no real 2.5h merge delay there. In fact, once the `sync` step is ran the merge factually happened.

True, didn't think of it that way.

There is still a way to make these things go faster: make them all run as part of the same stage, so you get parallelism among them - reducing the time to 45min if there are enough resources on the cluster.
I guess there are no actual dependencies between them?

**Comment by tcojean** [comment]
>There is still a way to make these things go faster: make them all run as part of the same stage, so you get parallelism among them - reducing the time to 45min if there are enough resources on the cluster. 
> I guess there are no actual dependencies between them?

That's a good idea. The only problem is that right now they use the same build directory, but I can find a way to change that in the script.

**Comment by tcojean** [comment]
One problem actually, if you guys know of any way to do that, is that we are running all of these steps, building, testing, coverage etc etc for **both** the private and the public repository. If there is a way to only run part of them once (such as coverage and friends) that would also be a great improvement.

**Comment by gflegar** [comment]
> The only problem is that right now they use the same build directory, but I can find a way to change that in the script.

Is that even an issue? I though that every job will get its own docker instance, with the build directories reset to the state you saved as the artifact of the job. But they will in fact be copied artifacts (since in reality, they can run on separate physical machines, depending on which runner is assigned that job).

> One problem actually, if you guys know of any way to do that, is that we are running all of these steps, building, testing, coverage etc etc for **both** the private and the public repository. If there is a way to only run part of them once (such as coverage and friends) that would also be a great improvement.

Look at the gh-pages stage, it's solved there: https://github.com/ginkgo-project/ginkgo/blob/develop/.gitlab-ci.yml#L333
Basically, my idea was to define a variable in the public CI that is not defined in the private one. This way, you can differentiate between the two. Actually, maybe we want to also do that for the build and test stages? As we're also running these twice. Then we do not get the correct checkmark for the private repository for these branches, but we can see that in the public one.

**Comment by tcojean** [comment]
>Is that even an issue?

Right, for some reason I initially thought as part of one job instead of multiple jobs in the same step...

> Basically, my idea was to define a variable in the public CI that is not defined in the private one. This way, you can differentiate between the two. Actually, maybe we want to also do that for the build and test stages? As we're also running these twice.

Thanks for the tip, looks good! I will look into that.
